exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1321 @@
+    // with the just-written batch.
+
+    String outputFile = spillSet.getNextSpillFile();
+    stats.setLongStat(Metric.SPILL_COUNT, spillSet.getFileCount());
+    BatchGroup.SpilledRun newGroup = null;
+    try (AutoCloseable ignored = AutoCloseables.all(batchesToSpill);
+         CopierHolder.BatchMerger merger = copierHolder.startMerge(schema, batchesToSpill, outputBatchRecordCount)) {
+      logger.trace("Merging and spilling to {}", outputFile);
+      newGroup = new BatchGroup.SpilledRun(spillSet, outputFile, oContext, spillSize);
+
+      // The copier will merge records from the buffered batches into
+      // the outputContainer up to targetRecordCount number of rows.
+      // The actual count may be less if fewer records are available.
+
+      while (merger.next()) {
+
+        // Add a new batch of records (given by merger.getOutput()) to the spill
+        // file, opening the file if not yet open, and creating the target
+        // directory if it does not yet exist.
+        //
+        // note that addBatch also clears the merger's output container
+
+        newGroup.addBatch(merger.getOutput());
+      }
+      injector.injectChecked(context.getExecutionControls(), INTERRUPTION_WHILE_SPILLING, IOException.class);
+      newGroup.closeOutputStream();
+      logger.trace("mergeAndSpill: completed, memory = {}, spilled {} records to {}",
+                   allocator.getAllocatedMemory(), merger.getRecordCount(), outputFile);
+      return newGroup;
+    } catch (Throwable e) {
+      // we only need to cleanup newGroup if spill failed
+      try {
+        if (newGroup != null) {
+          AutoCloseables.close(e, newGroup);
+        }
+      } catch (Throwable t) { /* close() may hit the same IO issue; just ignore */ }
+
+      // Here the merger is holding onto a partially-completed batch.
+      // It will release the memory in the close() call.
+
+      try {
+        // Rethrow so we can organize how to handle the error.
+
+        throw e;
+      }
+
+      // If error is a User Exception, just use as is.
+
+      catch (UserException ue) { throw ue; }
+      catch (Throwable ex) {
+        throw UserException.resourceError(ex)
+              .message("External Sort encountered an error while spilling to disk")
+              .build(logger);
+      }
+    }
+  }
+
+  /**
+   * Allocate and initialize the selection vector used as the sort index.
+   * Assumes that memory is available for the vector since memory management
+   * ensured space is available.
+   *
+   * @return a new, populated selection vector 2
+   */
+
+  private SelectionVector2 newSV2() {
+    SelectionVector2 sv2 = new SelectionVector2(allocator);
+    if (!sv2.allocateNewSafe(incoming.getRecordCount())) {
+      throw UserException.resourceError(new OutOfMemoryException("Unable to allocate sv2 buffer"))
+            .build(logger);
+    }
+    for (int i = 0; i < incoming.getRecordCount(); i++) {
```

Comment:
You're welcome.

Strangely, the `SingleBatchsorterTemplate` in the (unmanaged) external sort does not initialize the sv2.

Actually, this make sense. Notice that the code will reuse a selection vector from the upstream operator. In that case, we have to work with the sv values already applied and not try to recreate them as doing so will, say, unfilter the batch.